### ISSUE TITLE
Provide a method to migrate the dnf4 history database

### ISSUE THREAD

**Issue by jan-kolarik** [rpm-software-management/dnf5#1871]
dnf5 now automatically migrates the dnf4 database on first use. However, user reports (e.g. [here](https://github.com/rpm-software-management/dnf5/issues/1379) or [here](https://github.com/rpm-software-management/dnf5/discussions/1846)) indicate potential issues with incomplete or failed migrations.

To address this, it would be helpful to introduce an explicit method in dnf5 to manually trigger database migration from dnf4. 

Alternatively, dnf5 could perform a dnf4 database lookup whenever it encounters a missing state in its own database. If the latter approach is chosen, we should ensure it does not impact dnf5 performance and consider making it configurable.

**Comment by luksak** [comment]
Oh that feature would be great! Is there a way to replicate this manually until we have this feature?

**Comment by lantw44** [comment]
> dnf5 now automatically migrates the dnf4 database on first use.

Does DNF 5 really migrate things, including the history database? I installed a Fedora 40 VM, upgraded it to Fedora 41, and `dnf history` showed nothing. I hope there will be a tool which can make `dnf history` show the same thing as `dnf4 history` like how the YUM to DNF migration was done.

**Comment by kontura** [comment]
It is a little bit confusing, there are basically 2 separate issues here, dnf5 does migrate things from the history database but it migrates only system state information. It doesn't migrate transaction history which is what @lantw44 wants I think.

Perhaps we could implement the import of transaction history as well but we would have to figure out when to do it.

Regarding the issues in https://github.com/rpm-software-management/dnf5/issues/1379 I don't think repeated import would help much. dnf5 already got all the information it could about the system state from dnf4. See my latest comment there for some additional details.

**Comment by lantw44** [comment]
Yes, you are right. Actually I don't care the state too much since early DNF versions have reset installation reasons to unknown during system upgrades many times, and it is possible to backup installation reasons with `dnf repoquery` and restore them with `dnf mark`. The history database is much more valuable to me.

This missing feature in DNF 5 is the only reason I still use Fedora 40. Maybe I will be forced to upgrade to Fedora 41 soon, but it is likely I will keep using DNF 4 until the feature or a similar tool is available.

**Comment by kontura** [comment]
I have investigated the reports and I currently don't think manual import could help users.

Some of the difference are explained by https://github.com/rpm-software-management/dnf5/pull/1998.

For anyone else please report what is your problem, current state and how did you end up in it.
Don't hesitate to reopen this issue of create a new one.
